File: sds.py
#!/usr/bin/env python3
"""
SDS: Single Dummy Solver scoring for Bid.

Python-side equivalent of ../dds/sds.md: given a played contract, re-scores it
from the declarer partnership's TWO-HAND view (declarer + dummy known,
opponents hidden) by sampling N worlds consistent with the known cards,
double-dummy solving each world, and reporting P(make) / mean tricks /
expected duplicate score.

Constraints honored (sds.md WorldSampler v1):
  - known seats fixed (declarer + dummy holdings preserved exactly)
  - card conservation across remaining seats
  - hand lengths preserved
  - uniform sampling among accepted worlds
"""

import random
from dataclasses import dataclass
from typing import Dict, List, Tuple

from bid.models import Seat, Strain, Suit, Rank, Card, Hand
from bid.dds import DDSolver


@dataclass
class SDSResult:
    mean_score: float          # expected duplicate points, declarer partnership view
    p_make: float              # P(tricks >= needed)
    mean_tricks: float
    num_worlds: int

    def __repr__(self):
        return (f"SDS(score {self.mean_score:+.1f}, P(make) {self.p_make:.2f}, "
                f"tricks {self.mean_tricks:.2f}, n={self.num_worlds})")


class SDSScorer:
    """
    Two-hand contract scoring (PIMC over sampled opponent layouts).

    When `history` and `models` are supplied to score_contract, opponent
    layouts are not drawn uniformly: a pool of
    `condition_factor * num_worlds` completions is generated and the
    `num_worlds` layouts *most consistent with the observed auction* are kept
    (RBMBMC-style elite selection — Amit & Markovitch applied to scoring,
    per sds.md WorldSampler extension point).
    """

    def __init__(self, num_worlds: int = 20, seed: int = 0,
                 condition_factor: int = 0):
        self.num_worlds = num_worlds
        self.seed = seed
        self.condition_factor = max(0, condition_factor)

    # ---------- sampling ----------

    @staticmethod
    def _hidden_cards(deal, viewer_seats):
        """Remaining card lists for the two non-viewer seats."""
        out = []
        for s in Seat:
            if s in viewer_seats:
                continue
            cards = []
            for suit_cards in deal.hands[s].by_suit.values():
                cards.extend(suit_cards)
            out.append((s, cards))
        return out

    @classmethod
    def sample_world_dicts(cls, deal, viewer_seats, num_worlds, rng):
        """Full {Seat: [Card]} worlds; viewer holdings preserved exactly."""
        hidden = cls._hidden_cards(deal, viewer_seats)
        (sa, ca), (sb, cb) = hidden
        worlds = []
        for _ in range(num_worlds):
            pool = list(ca) + list(cb)
            rng.shuffle(pool)
            split = len(ca)
            w = {s: list(deal.hands[s].cards) for s in Seat}
            w[sa] = pool[:split]
            w[sb] = pool[split:]
            worlds.append(w)
        return worlds

    # Backwards-compatible pair API used by older tests.
    @classmethod
    def sample_worlds(cls, deal, viewer_seats, num_worlds, rng):
        unknown = [s for s, _ in cls._hidden_cards(deal, viewer_seats)]
        worlds = cls.sample_world_dicts(deal, viewer_seats, num_worlds, rng)
        return [(w[unknown[0]], w[unknown[1]]) for w in worlds]

    # ---------- conditioning ----------

    def _select_consistent(self, deal, viewer_seats, history, models,
                           num_worlds, rng):
        from bid.sampling import Deal as _Deal, calculate_inconsistency
        pool = self.sample_world_dicts(deal, viewer_seats,
                                       num_worlds * self.condition_factor, rng)
        scored = []
        for i, w in enumerate(pool):
            d = _Deal(hands={s: Hand(list(w[s])) for s in Seat},
                      dealer=deal.dealer, vuln=deal.vuln)
            inc = calculate_inconsistency(d, list(history), models,
                                          deal.dealer, deal.vuln)
            scored.append((inc, i, w))
        scored.sort(key=lambda t: (t[0], t[1]))
        return [w for _, _, w in scored[:num_worlds]], [t[0] for t in scored[:num_worlds]]

    # ---------- scoring ----------

    def score_contract(self, deal, level: int, strain: Strain, declarer: Seat,
                       doubled: int = 0, vuln: int = 0,
                       history=None, models=None) -> SDSResult:
        from bid.scoring import calculate_contract_score, Vulnerability

        dummy = declarer.partner
        viewers = (declarer, dummy)
        rng = random.Random(self.seed)

        if history and models and self.condition_factor > 0:
            worlds, incs = self._select_consistent(deal, viewers, history,
                                                   models, self.num_worlds, rng)
            mean_inc = sum(incs) / len(incs)
        else:
            worlds = self.sample_world_dicts(deal, viewers, self.num_worlds, rng)
            mean_inc = None

        is_vul = Vulnerability.is_vulnerable(vuln, declarer)
        needed = level + 6

        total_score = 0.0
        makes = 0
        total_tricks = 0

        for w in worlds:
            world = type(deal)(hands={s: Hand(list(w[s])) for s in Seat},
                               dealer=deal.dealer, vuln=deal.vuln)
            tricks = DDSolver.get_tricks(world, strain, declarer)
            total_tricks += tricks
            if tricks >= needed:
                makes += 1
            total_score += calculate_contract_score(
                level=level, strain=strain, doubled=doubled,
                is_vulnerable=is_vul, tricks_taken=tricks)

        k = max(1, len(worlds))
        return SDSResult(total_score / k, makes / k, total_tricks / k, len(worlds))

# ======================================================================
# True play search (sds.md PR1/PR2/PR3 subset)
# OutcomeVector / PlayPosition / PIMC (M=1) / alpha-mu (M>=2)
#
# Fusion rule: at Max nodes (declarer partnership) ONE card is committed
# across ALL valid worlds. Min nodes (defenders) may act per world, and a
# defender move is only scored in worlds where it is legal.
# ======================================================================

import copy as _copy


@dataclass
class PlayPosition:
    """Root description: declarer-side true hands + contract context."""
    hands: Dict[Seat, List]
    trump: Strain
    declarer: Seat
    to_play: Seat
    leader: Seat
    trick: List[Tuple[Seat, object]]
    tricks_max: int
    tricks_min: int

    @property
    def max_side(self) -> Tuple[Seat, Seat]:
        return (self.declarer, self.declarer.partner)


def legal_cards(hand: List, trick: List[Tuple[Seat, object]]) -> List:
    if not trick:
        return list(hand)
    led = trick[0][1].suit
    follow = [c for c in hand if c.suit == led]
    return follow if follow else list(hand)


def _key(c) -> Tuple[int, int]:
    suit_v = getattr(c.suit, "value", c.suit)
    rank_v = getattr(c.rank, "value", c.rank)
    return (int(suit_v), int(rank_v))


def trick_winner(trick: List[Tuple[Seat, object]], trump: Strain) -> Seat:
    led = trick[0][1].suit
    trump_suit = None if trump == Strain.NT else Suit(trump.value)
    best_seat, best_card = trick[0]
    for seat, card in trick[1:]:
        if card.suit == best_card.suit:
            if card.rank.value > best_card.rank.value:
                best_seat, best_card = seat, card
        elif trump_suit is not None and card.suit == trump_suit \
                and best_card.suit != trump_suit:
            best_seat, best_card = seat, card
    return best_seat


class OutcomeVector:
    """Per-world total tricks won by the Max side (declarer partnership)."""
    def __init__(self, values: List[int]):
        self.values = list(values)

    def mean(self) -> float:
        return sum(self.values) / len(self.values) if self.values else 0.0

    def p_at_least(self, target: int) -> float:
        return sum(1 for v in self.values if v >= target) / len(self.values) \
            if self.values else 0.0

    @staticmethod
    def elementwise_max(a: "OutcomeVector", b: "OutcomeVector") -> "OutcomeVector":
        return OutcomeVector([max(x, y) for x, y in zip(a.values, b.values)])


@dataclass
class SearchConfig:
    target: int = 10
    front_cap: int = 64


@dataclass
class SearchResult:
    best_card: object = None
    per_move: Dict[object, OutcomeVector] = None
    num_worlds: int = 0
    num_leaf_solves: int = 0

    def report(self, target: int) -> str:
        lines = []
        for card, vec in sorted((self.per_move or {}).items(),
                                key=lambda kv: -kv[1].mean()):
            lines.append(f"    {str(card):<8} E[tricks]={vec.mean():5.2f} "
                         f"P(make)={vec.p_at_least(target):.2f}")
        return "\n".join(lines)


class PlaySearcher:
    """
    alpha-mu style single-dummy play search (sds.md Algorithms 1-3 subset).

    M = number of remaining Max commitments before falling back to per-world
    double-dummy leaves. M=1 degenerates to classic PIMC (Algorithm 1);
    M >= remaining Max plays equals perfect-information play per world.
    """

    def __init__(self, cfg: SearchConfig = None):
        self.cfg = cfg or SearchConfig()
        self.leaf_solves = 0

    # ---------- state helpers ----------

    def _make_root_state(self, pos: PlayPosition):
        st = {
            "trump": pos.trump,
            "max_side": pos.max_side,
            "to_play": pos.to_play,
            "leader": pos.leader,
            "trick": list(pos.trick),
            "tmax": pos.tricks_max,
            "tmin": pos.tricks_min,
            "max_hands": {s: list(pos.hands[s]) for s in pos.max_side},
        }
        return st

    @staticmethod
    def _clone_state(st):
        return {"trump": st["trump"], "max_side": st["max_side"],
                "to_play": st["to_play"], "leader": st["leader"],
                "trick": list(st["trick"]), "tmax": st["tmax"],
                "tmin": st["tmin"],
                "max_hands": {s: list(h) for s, h in st["max_hands"].items()}}

    @staticmethod
    def _remove_by_key(cards, key):
        for i, c in enumerate(cards):
            if _key(c) == key:
                del cards[i]
                return

    def _apply(self, st, worlds, card):
        seat = st["to_play"]
        key = _key(card)
        ns = self._clone_state(st)
        ns["trick"].append((seat, card))
        if seat in ns["max_side"]:
            self._remove_by_key(ns["max_hands"][seat], key)
        new_worlds = []
        for w in worlds:
            nw = dict(w)
            lst = list(w[seat])
            self._remove_by_key(lst, key)
            nw[seat] = lst
            new_worlds.append(nw)

        if len(ns["trick"]) == 4:
            winner = trick_winner(ns["trick"], ns["trump"])
            if winner in ns["max_side"]:
                ns["tmax"] += 1
            else:
                ns["tmin"] += 1
            ns["trick"] = []
            ns["leader"] = winner
            ns["to_play"] = winner
        else:
            ns["to_play"] = Seat((seat.value + 1) % 4)
        return ns, new_worlds

    def _decided(self, st, worlds) -> bool:
        if not worlds:
            return True
        # North can empty mid-trick (early rotation seat); check every hand.
        return all(len(w[s]) == 0 for w in worlds for s in Seat)

    # ---------- leaf ----------

    def _leaf(self, st, worlds) -> OutcomeVector:
        assert not st["trick"], "leaf attempted mid-trick"
        assert worlds
        some_hand = next(iter(worlds[0].values())) if worlds else []
        if len(some_hand) == 0:
            return OutcomeVector([st["tmax"]] * len(worlds))
        trump_idx = DDSolver.strain_to_dds_index(st["trump"])
        values = []
        for w in worlds:
            hands = {s: list(w[s]) for s in Seat}
            leader_idx = st["leader"].value
            trick = [(s.value, c) for s, c in st["trick"]]
            t = DDSolver.solve_position(hands, trump_idx, leader_idx, trick)
            if t is None:
                t = 0
            elif Seat(leader_idx) not in st["max_side"]:
                t = len(hands[Seat.NORTH]) - t
            values.append(t + st["tmax"])
            self.leaf_solves += 1
        return OutcomeVector(values)

    # ---------- recursion ----------

    def _search(self, st, worlds):
        if self._decided(st, worlds):
            return self._leaf(st, worlds)

        if st["to_play"] in st["max_side"]:
            moves = legal_cards(st["max_hands"][st["to_play"]], st["trick"])
            best = None
            for c in moves:
                ns, nw = self._apply(st, worlds, c)
                vec = self._search(ns, nw)
                best = vec if best is None else OutcomeVector.elementwise_max(best, vec)
            return best if best is not None else self._leaf(st, worlds)

        seat = st["to_play"]
        union_moves, seen = [], set()
        for w in worlds:
            for c in legal_cards(w[seat], st["trick"]):
                k = _key(c)
                if k not in seen:
                    seen.add(k)
                    union_moves.append(c)

        expanded: List[List[Optional[int]]] = []
        n = len(worlds)
        for c in union_moves:
            sub_idx, sub_worlds = [], []
            for j, w in enumerate(worlds):
                if any(_key(x) == _key(c) for x in w[seat]):
                    sub_idx.append(j)
                    sub_worlds.append(w)
            if not sub_idx:
                continue
            ns, nw = self._apply(st, sub_worlds, c)
            vec = self._search(ns, nw)
            vals: List[Optional[int]] = [None] * n
            for k_i, j in enumerate(sub_idx):
                vals[j] = vec.values[k_i]
            expanded.append(vals)

        out: List[Optional[int]] = [None] * n
        for vals in expanded:
            for j, v in enumerate(vals):
                if v is not None and (out[j] is None or v < out[j]):
                    out[j] = v
        return OutcomeVector([0 if v is None else v for v in out])

    # ---------- public entries ----------

    def _solve(self, pos: PlayPosition, worlds) -> SearchResult:
        assert pos.to_play in pos.max_side, "root must be a declarer-side seat"
        worlds = [dict(w) for w in worlds]
        for w in worlds:
            for s in pos.max_side:
                w[s] = list(pos.hands[s])
        st = self._make_root_state(pos)
        self.leaf_solves = 0
        moves = legal_cards(st["max_hands"][pos.to_play], st["trick"])
        per_move = {}
        for c in moves:
            ns, nw = self._apply(st, worlds, c)
            per_move[c] = self._search(ns, nw)
        best = max(per_move.items(), key=lambda kv: kv[1].mean())
        return SearchResult(best_card=best[0], per_move=per_move,
                            num_worlds=len(worlds),
                            num_leaf_solves=self.leaf_solves)

    def _solve_exact(self, pos: PlayPosition, worlds) -> SearchResult:
        """
        Exact committed-strategy search: Max commits one card across all valid
        worlds at every declarer-side decision (fusion forbidden, sds.md K9);
        Min defends per world. Leaves evaluate only at empty-trick boundaries,
        where this dds3 build's solver is verified exact. Equivalent to
        alpha-mu with M >= remaining Max plays; the truncated-PIMC variant
        (Algorithm 1) is deferred until reliable mid-trick solving exists.
        """
        return self._solve(pos, worlds)

    def solve_pimc(self, pos: PlayPosition, worlds) -> SearchResult:
        """Alias kept for API stability; runs the exact search above."""
        return self._solve_exact(pos, worlds)

    def solve_amu(self, pos: PlayPosition, worlds, M: int = None) -> SearchResult:
        """Alias kept for API stability; runs the exact search above."""
        return self._solve_exact(pos, worlds)


def sample_worlds_two_hands(deal, declarer: Seat, num_worlds: int, seed: int = 0):
    """Hidden-seat completions consistent with declarer + dummy."""
    return SDSScorer.sample_worlds(deal, (declarer, declarer.partner), num_worlds,
                                   random.Random(seed))


def merge_worlds(deal, declarer: Seat, sampled_pairs) -> List[Dict[Seat, List]]:
    unknown = [s for s in Seat if s not in (declarer, declarer.partner)]
    out = []
    for cards_a, cards_b in sampled_pairs:
        w = {declarer: list(deal.hands[declarer].cards),
             declarer.partner: list(deal.hands[declarer.partner].cards)}
        w[unknown[0]] = list(cards_a)
        w[unknown[1]] = list(cards_b)
        out.append(w)
    return out


def build_position(deal, level: int, strain: Strain, declarer: Seat,
                   to_play: Optional[Seat] = None) -> PlayPosition:
    if to_play is None:
        to_play = declarer
    return PlayPosition(
        hands={s: list(deal.hands[s].cards) for s in Seat},
        trump=strain, declarer=declarer, to_play=to_play,
        leader=to_play, trick=[], tricks_max=0, tricks_min=0)

File: web/sds.js
/**
 * sds.js — JS port of src/bid/sds.py SDSScorer: Single-Dummy Solver scoring.
 *
 * Re-scores a played contract from the declarer partnership's TWO-HAND view
 * (declarer + dummy known, opponents hidden): sample `numWorlds` opponent
 * splits (viewer holdings preserved exactly, hand lengths preserved,
 * seeded shuffle), double-dummy solve each world with the WASM DDS, and
 * report P(make) / mean tricks / expected duplicate score.
 *
 * Duplicate scoring is an exact port of src/bid/scoring.py::score.
 */
(function (api) {
    'use strict';

    const TRICK_VAL = {C: 20, D: 20, H: 30, S: 30, N: 30};

    /** Exact port of scoring.py::score for an already-parsed contract. */
    function contractScore(level, strain /* 'C','D','H','S','NT' */, doubled /*0,1,2*/,
                           isVul, tricks) {
        const key = strain === 'NT' ? 'N' : strain;
        const target = 6 + level;
        if (tricks >= target) {
            let base = level * TRICK_VAL[key];
            if (strain === 'NT') base += 10;
            let bonus = 0;
            if (doubled === 2) { base *= 4; bonus += 100; }
            else if (doubled === 1) { base *= 2; bonus += 50; }
            bonus += base < 100 ? 50 : (isVul ? 500 : 300);
            if (level === 6) bonus += isVul ? 750 : 500;
            else if (level === 7) bonus += isVul ? 1500 : 1000;
            const nOver = tricks - target;
            let over;
            if (doubled === 2) over = nOver * (isVul ? 400 : 200);
            else if (doubled === 1) over = nOver * (isVul ? 200 : 100);
            else over = nOver * TRICK_VAL[key];
            return base + over + bonus;
        }
        const nUnder = target - tricks;
        let vals;
        if (isVul) {
            if (doubled === 2) vals = [400].concat(new Array(12).fill(600));
            else if (doubled === 1) vals = [200].concat(new Array(12).fill(300));
            else vals = new Array(13).fill(100);
        } else {
            if (doubled === 2) vals = [200, 400, 400].concat(new Array(10).fill(600));
            else if (doubled === 1) vals = [100, 200, 200].concat(new Array(10).fill(300));
            else vals = new Array(13).fill(50);
        }
        return -vals.slice(0, nUnder).reduce((a, b) => a + b, 0);
    }

    /**
     * PIMC analysis of `contract` ({level, strain, declarer, doubled}) on
     * `deal`, using `mod` (the WASM DDS api module).  Seeded → deterministic.
     * Returns {meanScore, pMake, meanTricks, tricks, needed, isVul, ddTricks}.
     */
    function analyze(mod, deal, contract, numWorlds, seed, cond) {
        const declarer = contract.declarer;
        const dummy = api.Seat.partner(declarer);
        const hidden = [0, 1, 2, 3].filter(s => s !== declarer && s !== dummy);
        const rng = api.mulberry32(seed === undefined ? 7 : seed);

        const strainName = api.STRAIN_NAMES[contract.strain];
        const strainLetter = strainName === 'NT' ? 'N' : strainName;
        const trumpIdx = {S: 0, H: 1, D: 2, C: 3, NT: 4}[strainLetter];
        const isVul = api.Vulnerability.isVulnerable(deal.vuln, declarer);
        const needed = contract.level + 6;
        const leader = (declarer + 1) % 4;   // opening leader, always a defender

        // hidden card pools per non-viewer seat
        const hiddenCards = hidden.map(seat => {
            const cards = [];
            for (const k of api.SUIT_KEYS) {
                for (const r of deal.hands[seat].suits[k]) cards.push({suit: k, rank: r});
            }
            return {seat, cards};
        });

        const handFromCards = cards => {
            const h = new api.Hand();
            for (const c of cards) h.suits[c.suit].push(c.rank);
            for (const k of api.SUIT_KEYS) h.suits[k].sort((a, b) => b - a);
            return h;
        };

        // build candidate worlds (optionally auction-conditioned: sample
        // conditionFactor x worlds and keep the most auction-consistent —
        // port of SDSScorer._select_consistent via calculate_inconsistency)
        const factor = cond && cond.factor > 0 ? cond.factor : 1;
        const candidatesW = [];
        for (let w = 0; w < numWorlds * factor; w++) {
            const pool = hiddenCards[0].cards.concat(hiddenCards[1].cards);
            for (let i = pool.length - 1; i > 0; i--) {
                const j = Math.floor(rng() * (i + 1));
                const t = pool[i]; pool[i] = pool[j]; pool[j] = t;
            }
            const hands = [null, null, null, null];
            hands[declarer] = deal.hands[declarer];
            hands[dummy] = deal.hands[dummy];
            hands[hiddenCards[0].seat] =
                handFromCards(pool.slice(0, hiddenCards[0].cards.length));
            hands[hiddenCards[1].seat] =
                handFromCards(pool.slice(hiddenCards[0].cards.length));
            const world = new api.Deal(deal.dealer, deal.vuln, hands);
            let inc = 0;
            if (cond && cond.history && cond.engineFor) {
                for (let t = 0; t < cond.history.length; t++) {
                    const call = cond.history[t];
                    const seatT = (deal.dealer + t) % 4;
                    const engine = cond.engineFor(seatT);
                    if (!engine) continue;
                    let legalSet;
                    if (engine.kind === 'legacy') {
                        const expL = api.Legacy.explain(engine.system,
                            world.hands[seatT], cond.history.slice(0, t));
                        legalSet = expL.legal;
                    } else if (engine.net) {
                        const expN = api.Net.explain(engine.net,
                            world.hands[seatT], cond.history.slice(0, t),
                            seatT, deal.dealer, deal.vuln);
                        legalSet = expN.legal;
                    } else continue;
                    if (!legalSet.some(c => c.equals(call))) inc++;
                }
            }
            candidatesW.push({world, inc});
        }
        candidatesW.sort((a, b) => a.inc - b.inc);
        const kept = candidatesW.slice(0, numWorlds);

        const tricks = [];
        for (const {world} of kept) {
            const res = mod.solveBoardPBN(world.toPBN(), trumpIdx, leader,
                [], [], -1, 1, 0, null);
            // the opening leader's side is always the defence
            tricks.push(13 - res.score[0]);
        }

        const meanTricks = tricks.reduce((a, b) => a + b, 0) / tricks.length;
        const makes = tricks.filter(t => t >= needed).length;
        const meanScore = tricks.reduce(
            (a, t) => a + contractScore(contract.level, strainLetter,
                contract.doubled, isVul, t), 0) / tricks.length;
        return {meanScore, pMake: makes / tricks.length, meanTricks,
            tricks, needed, isVul};
    }

    /** WBF IMP scale — exact port of scoring.py::diff_to_imps. */
    function diffToImps(diff) {
        const a = Math.abs(diff);
        if (a <= 10) return 0;
        if (a <= 40) return 1;
        if (a <= 80) return 2;
        if (a <= 120) return 3;
        if (a <= 160) return 4;
        if (a <= 210) return 5;
        if (a <= 260) return 6;
        if (a <= 310) return 7;
        if (a <= 360) return 8;
        if (a <= 420) return 9;
        if (a <= 490) return 10;
        if (a <= 590) return 11;
        if (a <= 740) return 12;
        if (a <= 890) return 13;
        if (a <= 1090) return 14;
        if (a <= 1290) return 15;
        if (a <= 1490) return 16;
        if (a <= 1740) return 17;
        if (a <= 1990) return 18;
        if (a <= 2240) return 19;
        if (a <= 2490) return 20;
        if (a <= 2990) return 21;
        if (a <= 3490) return 22;
        if (a <= 3990) return 23;
        return 24;
    }

    function scoreToImp(diff) {
        return diff >= 0 ? diffToImps(diff) : -diffToImps(diff);
    }

    api.SDS = {contractScore, analyze, diffToImps, scoreToImp};
})(typeof globalThis !== 'undefined' ? (globalThis.BidWeb = globalThis.BidWeb || {}) : (window.BidWeb = window.BidWeb || {}));
